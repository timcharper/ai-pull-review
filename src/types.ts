export type DIFF_STATUS = 'added' | 'modified' | 'removed';

export interface FileData {
  filename: string;
  change: string;
  originalContent: string;
  changes: string;
  status: DIFF_STATUS;
}

export interface AnalysisOptions {
  model: string;
  maxTokens: number;
  temperature?: number;
  maxRetries?: number;
  retryDelay?: number;
  commentThreshold?: number;
}

export interface GitHubConfig {
  token: string;
  repo: string;
  prNumber: number;
}

export interface LocalGitConfig {
  baseSha: string;
  headSha: string;
}

export interface Config {
  afterLines: number;
  beforeLines: number;
  anthropicApiKey: string;
  model: string;
  include: string[];
  exclude: string[];
  maxFiles: number;
  maxSize: number;
  commentThreshold: number;
  writePullRequest: boolean;
  output: string;
  dryRun: boolean;
  workingDir: string;
  mode: 'github' | 'localgit';
  github?: GitHubConfig;
  localgit?: LocalGitConfig;
}

export interface FileObject {
  filename: string;
  status: string;
  additions: number;
  deletions: number;
  changes: number;
  patch?: string;
}

export interface FilterOptions {
  includePatterns: string[];
  excludePatterns: string[];
  maxFiles: number;
}

export interface Message {
  role: MessageRole;
  content: string;
}

export interface BatchResult {
  fileName: string;
  content: string;
}

export interface ModelCost {
  model: string;
  cost: number;
}

export type MessageRole = 'user' | 'assistant';

export interface MessageParam {
  role: MessageRole;
  content: string;
}
