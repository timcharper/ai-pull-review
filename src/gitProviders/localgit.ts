import { execSync } from 'child_process';
import { GitProvider, DiffSet, DiffSetEntry } from '../gitProvider';
import { Config, LocalGitConfig, DIFF_STATUS, CommitMessage } from '../types';
import { minimatch } from 'minimatch';
import assert from 'assert';
import { logger } from '../utils/logger';
import { filterPathsWithGlobAndPrefix, globToRegex } from '../utils/glob';

export function inferMainlineBranch(cwd: string): string {
  try {
    // Try to get the default branch from origin/HEAD
    const defaultBranch = execSync('git symbolic-ref refs/remotes/origin/HEAD', {
      encoding: 'utf8',
      cwd,
    })
      .trim()
      .replace('refs/remotes/origin/', '');
    return defaultBranch;
  } catch (error) {
    // Fallback to common branch names
    const commonBranches = ['main', 'master'];
    for (const branch of commonBranches) {
      try {
        execSync(`git show-ref --verify --quiet refs/remotes/origin/${branch}`, { cwd });
        return branch;
      } catch {
        continue;
      }
    }
    // If all else fails, use HEAD~1 as a safe default
    return 'HEAD~1';
  }
}

export class LocalGitDiffProvider implements GitProvider {
  constructor(
    private config: Config,
    private localgitConfig: LocalGitConfig,
  ) {}

  static async getInstance(config: Config): Promise<LocalGitDiffProvider> {
    assert(config.localgit, 'LocalGit config is required');
    return new LocalGitDiffProvider(config, config.localgit);
  }

  async getDiff(): Promise<DiffSet> {
    const { baseSha, headSha } = this.localgitConfig;

    // Get the diff between the two commits
    const cmd = `git diff ${baseSha}...${headSha}`;
    logger.debug('Running command:', cmd);
    const diffOutput = execSync(cmd, {
      encoding: 'utf8',
      cwd: this.config.workingDir,
    });

    // Parse the diff output into files
    const files: DiffSetEntry[] = [];
    let currentFile: Partial<DiffSetEntry> = {};

    for (const line of diffOutput.split('\n')) {
      if (line.startsWith('diff --git')) {
        if (currentFile.filename) {
          files.push(currentFile as DiffSetEntry);
        }
        const parts = line.split(' ');
        const filepath = parts[2]?.replace('a/', '');
        if (filepath) {
          currentFile = {
            filename: filepath,
            status: 'modified',
            patch: '',
          };
        }
      } else if (currentFile.filename) {
        currentFile.patch = (currentFile.patch || '') + line + '\n';
      }
    }

    if (currentFile.filename) {
      files.push(currentFile as DiffSetEntry);
    }

    return {
      files,
      baseSha,
      headSha,
    };
  }

  async writeComment(comment: string): Promise<void> {
    // For local Git, we'll just print the comment to stdout
    console.log('=== Analysis Results ===');
    console.log(comment);
    console.log('=======================');
  }

  async getFileContent(filename: string): Promise<string> {
    try {
      return execSync(`git show ${this.localgitConfig.headSha}:${filename}`, {
        encoding: 'utf8',
        cwd: this.config.workingDir,
      });
    } catch (error) {
      console.error(`Error getting file content for ${filename}:`, error);
      return '';
    }
  }

  async listFiles(path: string, glob: string): Promise<string[]> {
    try {
      // List all files in the repo
      const allFiles = execSync('git ls-files', {
        encoding: 'utf8',
        cwd: this.config.workingDir,
      })
        .split('\n')
        .filter(Boolean);

      return filterPathsWithGlobAndPrefix(allFiles, path, glob);
    } catch (error) {
      logger.error(`Error listing files in ${path} matching ${glob}:`, error);
      return [];
    }
  }

  async getCommitMessages(baseSha: string, headSha: string): Promise<CommitMessage[]> {
    try {
      const cmd = `git log ${baseSha}..${headSha} --pretty=format:"%H|%an|%ad|%s" --date=iso`;
      logger.debug('Getting commit messages with command:', cmd);

      const output = execSync(cmd, {
        encoding: 'utf8',
        cwd: this.config.workingDir,
      });

      if (!output.trim()) {
        return [];
      }

      const commits: CommitMessage[] = output
        .split('\n')
        .filter(Boolean)
        .map((line) => {
          const [sha, author, date, ...messageParts] = line.split('|');
          return {
            sha: sha || '',
            author: author || '',
            date: date || '',
            message: messageParts.join('|') || '',
          };
        });

      logger.debug('Found commit messages:', commits.length);
      return commits;
    } catch (error) {
      logger.error('Error getting commit messages:', error);
      return [];
    }
  }
}
