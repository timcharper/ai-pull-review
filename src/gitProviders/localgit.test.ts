import { LocalGitDiffProvider } from './localgit';
import { Config } from '../types';
import path from 'path';
import assert from 'assert';

describe('LocalGitDiffProvider', () => {
  const repoPath = path.resolve(__dirname, '../../test/fixtures/test-git-repo');
  const config: Config = {
    anthropicApiKey: 'test-key',
    model: 'test-model',
    analysisLevel: 'basic',
    include: ['*.txt', '*.md'],
    exclude: [],
    maxFiles: 100,
    maxSize: 1000,
    commentThreshold: 0.5,
    writePullRequest: false,
    output: 'stdout',
    workingDir: repoPath,
    mode: 'localgit',
    localgit: {
      baseSha: 'main', // Initial commit
      headSha: 'HEAD', // Feature branch
    },
  };

  let provider: LocalGitDiffProvider;

  beforeAll(async () => {
    provider = await LocalGitDiffProvider.getInstance(config);
  });

  describe('listFiles', () => {
    it('should list all files matching glob pattern', async () => {
      const files = await provider.listFiles('', '*.txt');
      expect(files).toContain('feature.txt');
    });

    it('should list files in specific directory', async () => {
      const files = await provider.listFiles('', '*.md');
      expect(files).toContain('README.md');
    });

    it('should return empty array for non-matching pattern', async () => {
      const files = await provider.listFiles('', '*.nonexistent');
      expect(files).toHaveLength(0);
    });

    it('should find files using recursive glob patterns', async () => {
      const files = await provider.listFiles('', '**/*.mdc');
      console.log('Files found:', files);
      expect(files).toContain('.cursor/rules/codes-files.mdc');
      expect(files).toContain('.cursor/rules/text-files.mdc');
    });
  });

  describe('getDiff', () => {
    it('should get diff between commits', async () => {
      const diff = await provider.getDiff();
      expect(diff.baseSha).toBe('main');
      expect(diff.headSha).toBe('HEAD');
      expect(diff.files).toHaveLength(4);
      const featureChange = diff.files.find((file) => file.filename === 'feature.txt');
      assert(featureChange, 'should be a feature change');
      expect(featureChange.status).toBe('modified');
      expect(featureChange.patch).toContain('Feature branch');
    });
  });

  describe('getFileContent', () => {
    it('should get content of existing file', async () => {
      const content = await provider.getFileContent('feature.txt');
      expect(content).toContain('Feature branch');
    });

    it('should return empty string for non-existent file', async () => {
      const content = await provider.getFileContent('nonexistent.txt');
      expect(content).toBe('');
    });
  });

  describe('writeComment', () => {
    it('should write comment to stdout', async () => {
      const consoleSpy = jest.spyOn(console, 'log');
      await provider.writeComment('Test comment');
      expect(consoleSpy).toHaveBeenCalledWith('=== Analysis Results ===');
      expect(consoleSpy).toHaveBeenCalledWith('Test comment');
      expect(consoleSpy).toHaveBeenCalledWith('=======================');
      consoleSpy.mockRestore();
    });
  });
});
