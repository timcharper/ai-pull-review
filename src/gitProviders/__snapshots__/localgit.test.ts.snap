// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalGitDiffProvider getCommitMessages should get commit messages 1`] = `
[
  {
    "author": "Tim Harper",
    "date": "2025-07-16 13:28:16 -0600",
    "message": "Add cursor rules",
    "sha": "8f22ee5d6658b63f18122afb3a4028b805428127",
  },
  {
    "author": "Tim Harper",
    "date": "2025-07-16 13:28:15 -0600",
    "message": "Feature branch",
    "sha": "8c776a3c913e0abdbe5840a8ff8d483ec2cb5805",
  },
]
`;
